ascon-hash/src/lib.rs

```diff
@@ -0,0 +1,462 @@
+            phantom: PhantomData,
+        }
+    }
+}
+
+/// Ascon hash implementation
+#[derive(Clone, Debug, Default)]
+pub struct AsconCore {
+    state: HashCore<Parameters>,
+}
+
+impl HashMarker for AsconCore {}
+
+impl BlockSizeUser for AsconCore {
+    type BlockSize = U8;
+}
+
+impl BufferKindUser for AsconCore {
+    type BufferKind = Eager;
+}
+
+impl OutputSizeUser for AsconCore {
+    type OutputSize = U32;
+}
+
+impl UpdateCore for AsconCore {
+    fn update_blocks(&mut self, blocks: &[Block<Self>]) {
+        for block in blocks {
+            self.state.absorb_block(block.as_ref());
+        }
+    }
+}
+
+impl FixedOutputCore for AsconCore {
+    fn finalize_fixed_core(&mut self, buffer: &mut Buffer<Self>, out: &mut Output<Self>) {
+        debug_assert!(buffer.get_pos() < 8);
+        self.state
+            .absorb_last_block(&buffer.get_data()[..buffer.get_pos()]);
+        self.state.squeeze(out);
+    }
+}
+
+impl Reset for AsconCore {
+    fn reset(&mut self) {
+        *self = Default::default();
+    }
+}
+
+impl AlgorithmName for AsconCore {
+    fn write_alg_name(f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        f.write_str("AsconHash")
+    }
+}
+
+/// Ascon hash implementation
+#[derive(Clone, Debug, Default)]
+pub struct AsconACore {
+    state: HashCore<ParametersA>,
+}
+
+impl HashMarker for AsconACore {}
+
+impl BlockSizeUser for AsconACore {
+    type BlockSize = U8;
+}
+
+impl BufferKindUser for AsconACore {
+    type BufferKind = Eager;
+}
+
+impl OutputSizeUser for AsconACore {
+    type OutputSize = U32;
+}
+
+impl UpdateCore for AsconACore {
+    fn update_blocks(&mut self, blocks: &[Block<Self>]) {
+        for block in blocks {
+            self.state.absorb_block(block.as_ref());
+        }
+    }
+}
+
+impl FixedOutputCore for AsconACore {
+    fn finalize_fixed_core(&mut self, buffer: &mut Buffer<Self>, out: &mut Output<Self>) {
+        debug_assert!(buffer.get_pos() < 8);
+        self.state
+            .absorb_last_block(&buffer.get_data()[..buffer.get_pos()]);
+        self.state.squeeze(out);
+    }
+}
+
+impl Reset for AsconACore {
+    fn reset(&mut self) {
+        *self = Default::default();
+    }
+}
+
+impl AlgorithmName for AsconACore {
+    fn write_alg_name(f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        f.write_str("AsconAHash")
+    }
+}
+
+/// Ascon XOF
+#[derive(Clone, Debug, Default)]
+pub struct AsconXOFCore {
+    state: HashCore<ParametersXOF>,
+}
+
+impl HashMarker for AsconXOFCore {}
+
+impl BlockSizeUser for AsconXOFCore {
+    type BlockSize = U8;
+}
+
+impl BufferKindUser for AsconXOFCore {
+    type BufferKind = Eager;
+}
+
+impl UpdateCore for AsconXOFCore {
+    fn update_blocks(&mut self, blocks: &[Block<Self>]) {
+        for block in blocks {
+            self.state.absorb_block(block.as_ref());
+        }
+    }
+}
+
+/// Reader for XOF output
+#[derive(Clone, Debug)]
+pub struct AsconXOFReaderCore {
```

Comment:
To follow the UpperCamelCase naming convention:

```suggestion
pub struct AsconXofReaderCore {
```

This also better matches the [`digest::XofReader`](https://docs.rs/digest/latest/digest/trait.XofReader.html) trait